grafana.ini yaml syntax

intermittentnrg opened the pull request:
Untested draft  for early feedback of #210.

 I've done search&replace to use new convention for `grafana_ini:` var.
I did not expect there to be 19 changed files, but there were many references for example to `grafana_url` which is now `grafana_ini.server.root_url` and similar.

I kept all entries in README and defaults/main.yaml, but I think most should be deleted. Repeating grafana documentation and default values doesn't make sense. The [default grafana.ini is empty/everything commented out](https://github.com/grafana/grafana/blob/main/conf/sample.ini), we should let grafana handle defaults.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana-ansible-collection?pullRequest=232) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana-ansible-collection?pullRequest=232) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana-ansible-collection?pullRequest=232) it.</sub>

intermittentnrg commented:
I reviewed all vars to discover which are not used in role and can be descoped/deferred to grafana docs.

Vars used in role:
| old | new | purpose |
| --- | --- | ------- |
| grafana_instance | grafana_ini.instance_name | default value from ansible |
| grafana_logs_dir | grafana_ini.paths.logs | mkdir |
| grafana_data_dir | grafana_ini.paths.data | provisioning dashboards & plugins |
| grafana_address | grafana_ini.server.http_addr | default for root_url |
| grafana_port | grafana_ini.server.http_port | default for root_url and validate cap_net_bind_service |
| grafana_url | grafan_ini.server.root_url | default for grafana_api_url |
| grafana_domain | grafana_ini.server.domain | default value from ansible |
| grafana_server | grafana_ini.server | create socket if protocol == 'socket |
| grafana_security | grafana_ini.security | admin_user and admin_password for api calls |
| grafana_database | grafana_ini.database | validate correct keys for type =/!=sqlite3 |
| grafana_auth | grafana_ini.auth | configure ldap if ldap.enabled |
| grafana_users | grafana_ini.users | allow_sign_up=false and default_theme=dark differ from grafana defaults |

Vars not used in role - removed from README.md and defaults/main.yaml in 2nd commit
| old | new | note |
| --- | --- | ---- |
| grafana_users | grafana_ini.users |
| grafana_welcome_email_on_sign_up | grafana_ini.emails.welcome_email_on_sign_up |
| grafana_session | grafana_ini.session |
| grafana_analytics | grafana_ini.analytics |
| grafana_smtp | grafana_ini.smtp |
| grafana_alerting | grafana_ini.alerting |
| grafana_log | grafana_ini.log |
| grafana_metrics | grafana_ini.metrics |
| grafana_tracing | grafana_ini."tracing.jaeger" |
| grafana_snapshots | grafana_ini.snapshots |
| grafana_image_storage | grafana_ini.image_storage |
| grafana_date_formats | grafana_ini.date_formats |
| grafana_feature_toggles | grafana_ini.feature_toggles |
| grafana_remote_cache | grafana_ini.remote_cache | undocumented |
| grafana_auth_generic_oauth | grafana_ini."auth.generic_oauth" | undocumented |
| grafana_panels | grafana_ini.panels | undocumented, no defaults |

Removed hardcoded sections in grafana.ini.j2 (could submit these as a separate PR, everything else is intertwined)
| section | note |
| ------- | ---- |
| [dashboards] | hardcoded, grafana defaults |
| [dashboards.json] | hardcoded, replaced by dashboards provisioning in grafana 5.0+ |
| [grafana_com] | hardcoded, grafana defaults |

intermittentnrg commented:
Finally got around to testing this.

The dict grafana_ini gets overwritten and not inherited from defaults/main.yaml.
I renamed it to `grafana_ini_default` and added this task in tasks/main.yaml:
```yaml
ansible.builtin.set_fact:
  grafana_ini: "{{ grafana_ini_default | ansible.builtin.combine(grafana_ini, recursive=true) }}"
```

Also this line no longer works as it refers to itself. So just set this explictly.
http_addr defaults to 0.0.0.0 which is not valid.
```yaml
grafana_ini:
  server:
    root_url: "http://{{ grafana_ini.server.http_addr }}:{{ grafana_ini.server.http_port }}"
```

I rebased and force pushed to fix conflicts.

intermittentnrg commented:
Awesome, thanks! I'm not sure how to run tests/molecule and just made the same changes that made sense elsewhere. If someone can confirm they're good or point me to how they are run.

intermittentnrg commented:
I rebased and force pushed.

How do I bump major/minor version and where do I add upgrade notes? Seems changelog is updated automatically from commit messages?

intermittentnrg commented:
Hmm I'm considering forking and publishing to ansible-galaxy as this PR is not receiving feedback or getting merged fast. Then I can use it properly for my own purposes.

intermittentnrg commented:
Awesome!

I realized this command installs collection from my git without publishing to ansible galaxy
```
ansible-galaxy collection install git+https://github.com/intermittentnrg/grafana-ansible-collection.git,grafana_ini
```

Also I was trying to make a preflight check for nested dicts.
Bad example:
```yaml
grafana_ini:
  auth:
    anonymous:
      enabled: true
```
Good example:
```yaml
grafana_ini:
  auth.anonymous:
    enabled: true
```

It's a struggle looping over nested dicts in ansible but I came up with this:
```yaml

- name: No nested mappings
  set_fact:
    grafana_ini_check: |-
      {% for section, items in grafana_ini.items() %}
      {% if items is mapping %}
      {% for k,v in items.items() %}
      {% if v is mapping %}
      grafana_ini.{{section}}.{{k}} can't be a dict!!
      {% endif %}
      {% endfor %}
      {% endif %}
      {% endfor %}
- name: No nested mappings
  ansible.builtin.assert:
    that: grafana_ini_check | length == 0
    fail_msg: "{{ grafana_ini_check }}"
```

Not sure this preflight check is necessary, but I just made above error and only noticed when I checked /etc/grafana/grafana.ini on target host.
Could check and compare with how grafana helm chart for kubernetes handles this? But anyhow this check is not strictly necessary.
Thought I'd at least leave the snippet it here if it turns out to be useful.
